identity-android-csa/src/main/java/com/android/identity/android/securearea/cloud/CloudSecureArea.kt

```diff
@@ -0,0 +1,893 @@
+                    .add(deviceNonce)
+                    .end()
+                    .build()
+            )
+            if (!Crypto.checkSignature(
+                    cloudBindingKey!!,
+                    dataSignedByTheCloud,
+                    Algorithm.ES256,
+                    response1.signature
+                )
+            ) {
+                throw CloudException("Error verifying signature")
+            }
+
+            // Now we can derive SKDevice and SKCloud
+            val Zab = Crypto.keyAgreement(eDeviceKey, response1.eCloudKey.ecPublicKey)
+            val salt = Crypto.digest(
+                Algorithm.SHA256,
+                Cbor.encode(
+                    CborArray.builder()
+                        .add(deviceNonce)
+                        .add(response0.cloudNonce)
+                        .end()
+                        .build()
+                )
+            )
+            skDevice = Crypto.hkdf(
+                Algorithm.HMAC_SHA256,
+                Zab,
+                salt,
+                "SKDevice".toByteArray(),
+                32
+            )
+            skCloud = Crypto.hkdf(
+                Algorithm.HMAC_SHA256,
+                Zab,
+                salt,
+                "SKCloud".toByteArray(),
+                32
+            )
+            e2eeContext = response1.serverState
+            encryptedCounter = 1
+            decryptedCounter = 1
+        } catch (e: Exception) {
+            throw CloudException(e)
+        }
+    }
+
+    private fun encryptToCloud(messagePlaintext: ByteArray): ByteArray {
+        // The IV and these constants are specified in ISO/IEC 18013-5:2021 clause 9.1.1.5.
+        val iv = ByteBuffer.allocate(12)
+        iv.putInt(0, 0x00000000)
+        val ivIdentifier = 0x00000001
+        iv.putInt(4, ivIdentifier)
+        iv.putInt(8, encryptedCounter++)
+        return Crypto.encrypt(Algorithm.A128GCM, skDevice!!, iv.array(), messagePlaintext)
+    }
+
+    private fun decryptFromCloud(messageCiphertext: ByteArray): ByteArray {
+        val iv = ByteBuffer.allocate(12)
+        iv.putInt(0, 0x00000000)
+        val ivIdentifier = 0x00000000
+        iv.putInt(4, ivIdentifier)
+        iv.putInt(8, decryptedCounter++)
+        return Crypto.decrypt(Algorithm.A128GCM, skCloud!!, iv.array(), messageCiphertext)
+    }
+
+    // This is internal rather than private b/c it's used in testPassphraseCannotBeChanged()
+    internal fun communicateE2EE(requestData: ByteArray): ByteArray {
+        return communicateE2EEInternal(requestData, 0)
+    }
+
+    private fun communicateE2EEInternal(
+        requestData: ByteArray,
+        callDepth: Int
+    ): ByteArray {
+        val requestWrapper = E2EERequest(
+            encryptToCloud(requestData),
+            e2eeContext!!
+        )
+        val (httpResponseCode, response) = runBlocking {
+            communicateLowlevel(serverUrl, requestWrapper.toCbor())
```

Comment:
This is, of course, undesirable; not for this PR, but can we aim to make all relevant API suspend instead?

Reply:
Yes, this is indeed the plan!